[JSC] Fix VMTraps race condition in fireTrap() for async events

hyjorc1 opened the pull request:
#### 2a837da2d80e4bf5745127d9e425524a189207c1
<pre>
[JSC] Fix VMTraps race condition in fireTrap() for async events
<a href="https://bugs.webkit.org/show_bug.cgi?id=302605">https://bugs.webkit.org/show_bug.cgi?id=302605</a>
<a href="https://rdar.apple.com/164842717">rdar://164842717</a>

Reviewed by Mark Lam and Dan Hecht.

There was a race condition where m_trapBits and m_threadStopRequested could
be observed in an inconsistent state, causing assertion failures.

The bug occurred in VMTraps::fireTrap() for async events:
1. Thread A calls fireTrap() and executes m_trapBits.exchangeOr(event)
   atomically WITHOUT holding m_trapSignalingLock.
2. Thread B in cancelThreadStopIfNeeded() acquires m_trapSignalingLock first.
3. Thread B reads m_trapBits and sees the bit set by Thread A.
4. Thread B checks m_threadStopRequested, which is still false because Thread A
   hasn&apos;t acquired the lock yet to call requestThreadStopIfNeeded()
5. This breaks the assertion in cancelThreadStopIfNeeded() that
   m_threadStopRequested must be true when needHandling(AsyncEvents) is true.

The fix introduces updateThreadStopRequestIfNeeded(), which synchronizes
m_threadStopRequested to match m_trapBits regardless of how we got to the
current state. Instead of having separate requestThreadStopIfNeeded() and
cancelThreadStopIfNeeded() calls that must reason about complex interleavings,
we have a single function that ensures the flag matches the source of truth.

Key changes:
- fireTrap() and clearTrap() now call updateThreadStopRequestIfNeeded() after
  modifying trap bits (with compiler fence to prevent reordering).
- updateThreadStopRequestIfNeeded() checks if m_threadStopRequested matches
  needHandling(AsyncEvents) and calls the appropriate helper if out of sync.
- requestThreadStopIfNeeded() and cancelThreadStopIfNeeded() are simplified
  to just perform their work, with asserts instead of guards.
- All interleavings now converge to the correct state because we sync to
  the source of truth rather than trying to maintain invariants.

This approach is conceptually simpler: instead of reasoning about whether to
request or cancel based on what operation we&apos;re performing, we simply ensure
the derived state (m_threadStopRequested) matches the source of truth
(m_trapBits). The lock-free fast path for trap bit modifications is preserved.

Tests: The corresponding stress tests will be attached in a follow-up patch
Canonical link: <a href="https://commits.webkit.org/303185@main">https://commits.webkit.org/303185@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b4db8bf0d186d4f8c4ff2811849bede133a15ed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131625 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3957 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42591 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139134 "Failed to checkout and rebase branch from PR 54019") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3989 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3798 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/139134 "Failed to checkout and rebase branch from PR 54019") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134571 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/117813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/139134 "Failed to checkout and rebase branch from PR 54019") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82325 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/123650 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/111397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/35939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141779 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130079 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3701 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/36455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/108866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3749 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/109110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/2804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/114142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56911 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20448 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3762 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/32543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/163097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3589 "Failed to checkout and rebase branch from PR 54019") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67173 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/163097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/3844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/3692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/JavaScriptCore/runtime/VMTraps.cpp:
```diff
@@ -399,9 +399,9 @@ CONCURRENT_SAFE void VMTraps::cancelThreadStopIfNeeded()
 }
 
-CONCURRENT_SAFE void VMTraps::requestThreadStopIfNeeded(VMTraps::Event event)
+CONCURRENT_SAFE void VMTraps::requestThreadStopIfNeeded(Locker<Lock>& locker, VMTraps::Event event)
```
`AbstractLocker&`, and leave out the name of the parameter. Then you don't need `UNUSED_PARAM`.

hyjorc1 replied:
We need the locker in the code below:
```
#if ENABLE(SIGNAL_BASED_VM_TRAPS)
    if (!Options::usePollingTraps()) {
        // sendSignal() can loop until it has confirmation that the mutator thread
        // has received the trap request. We'll call it from another thread so that
        // requestThreadStopIfNeeded() does not block.
        if (!m_signalSender)
            m_signalSender = adoptRef(new SignalSender(locker, vm));
        m_signalSender->notify(locker);
    }
#endif
```

dhecht replied:
Ah, didn't expand the diff enough, my bad.

MenloDorian commented on Source/JavaScriptCore/runtime/VMTraps.h:
```diff
@@ -244,9 +244,15 @@ class VMTraps {
+
         m_trapBits.exchangeOr(event);
-        if (isAsyncEvent(event))
-            requestThreadStopIfNeeded(event);
```
This is incorrect.  `m_trapBits` must always be set / cleared before we request / cancel the stop.  Otherwise, a race can occur where the handling of the stop occurs but cannot see the requested trap bit yet.  `m_trapBits` is the ultimate source of truth.  We just have to make sure that the stop request is not still installed when we no longer need to stop.

So, here's what we should do:
1. In `clearTrap()`, add a `compilerFence()` after `clearTrapWithoutCancellingThreadStop()` to document that the trap bit must be cleared first and cannot be moved below there.  Maybe add a comment also if you think the compiler fence is insufficient to document this.

2. In `fireTrap()`, add a `compilerFence()` after the setting of `m_trapBits` to document that the trap bit must be set first and cannot be moved below there.  Ditto about the comment.

3.  In `VMTraps::cancelThreadStopIfNeeded()`, 
     3.1 Move the `if (!m_threadStopRequested)` to the top of the function (after the lock).  If the stop is already cancelled, there's nothing more to do.  Add a `compilerFence()` after this if check and document that this must be checked before the `needHandling` check.
     3.2 Remove the `RELEASE_ASSERT` in the `if (needHandling(AsyncEvents))` check.  Now, we can only get there if `m_threadStopRequested` is true.

4. In `VMTraps::requestThreadStopIfNeeded()`,
    4.1 Move the `if (m_threadStopRequested)` to the top of the function (after the lock).  If the stop is already requested, there's nothing more to do.  Add a `compilerFence()` after this if check and document that this must be checked before the `needHandling` check.
    4.2 Remove the `RELEASE_ASSERT` in the `if (!needHandling(AsyncEvents))` check.  Now, we can only get there if `m_threadStopRequested` is false.

hyjorc1 replied:
Discussed offline with Mark and Dan. We should introduces `updateThreadStopRequestIfNeeded()`, which synchronizes `m_threadStopRequested` to match `m_trapBits` regardless of the current state.

MenloDorian commented on Source/JavaScriptCore/runtime/VMTraps.h:
```diff
@@ -293,8 +295,9 @@ class VMTraps {
-    JS_EXPORT_PRIVATE CONCURRENT_SAFE void requestThreadStopIfNeeded(Event);
+    CONCURRENT_SAFE void cancelThreadStopIfNeeded() WTF_REQUIRES_LOCK(m_trapSignalingLock);
+    CONCURRENT_SAFE void requestThreadStopIfNeeded(Locker<Lock>&) WTF_REQUIRES_LOCK(m_trapSignalingLock);
+    JS_EXPORT_PRIVATE CONCURRENT_SAFE void updateThreadStateIfNeeded();
```
nit: Please also rename updateThreadStateIfNeeded to updateThreadStopStateIfNeeded or updateThreadStopMachineryIfNeeded or just updateThreadStopIfNeeded (to be consistent with cancelThreadStopIfNeeded and requestThreadStopIfNeeded). I think updateThreadStateIfNeeded is too general and implies that it modifies other types of thread state that it does not (e.g. native thread suspension, or register / stack state).

hyjorc1 replied:
Renamed to `updateThreadStopStateIfNeeded `

MenloDorian commented on Source/JavaScriptCore/runtime/VMTraps.cpp:
```diff
@@ -380,41 +380,19 @@ void VMTraps::willDestroyVM()
-    if (!m_threadStopRequested)
-        return; // Cancel was already processed due to another thread. Nothing more to do.
+    ASSERT(m_threadStopRequested);
+    ASSERT(!needHandling(AsyncEvents));
```
I don't think you can assert this because another thread may have raced you and set the bit before you got into this function.  Remove this ASSERT.

MenloDorian commented on Source/JavaScriptCore/runtime/VMTraps.cpp:
```diff
@@ -380,41 +380,19 @@ void VMTraps::willDestroyVM()
 {
-    Locker locker { *m_trapSignalingLock };
+    ASSERT(!m_threadStopRequested);
+    ASSERT(needHandling(AsyncEvents));
```
I don't think you can assert this because another thread may have raced you and cleared the bit before you got into this function.  Remove this ASSERT.

MenloDorian reviewed (approved):
r=me with fixes.

Can you update note #5 in the commit message to say this instead: "This breaks the assertion in cancelThreadStopIfNeeded() that m_threadStopRequested must be true when needHandling(AsyncEvents) is true."  Calling it an invariant suggests that this assertion must be invariantly true when it is not.

hyjorc1 committed:
[JSC] Fix VMTraps race condition in fireTrap() for async events
https://bugs.webkit.org/show_bug.cgi?id=302605
rdar://164842717

Reviewed by Mark Lam and Dan Hecht.

There was a race condition where m_trapBits and m_threadStopRequested could
be observed in an inconsistent state, causing assertion failures.

The bug occurred in VMTraps::fireTrap() for async events:
1. Thread A calls fireTrap() and executes m_trapBits.exchangeOr(event)
   atomically WITHOUT holding m_trapSignalingLock.
2. Thread B in cancelThreadStopIfNeeded() acquires m_trapSignalingLock first.
3. Thread B reads m_trapBits and sees the bit set by Thread A.
4. Thread B checks m_threadStopRequested, which is still false because Thread A
   hasn't acquired the lock yet to call requestThreadStopIfNeeded()
5. This breaks the assertion in cancelThreadStopIfNeeded() that
   m_threadStopRequested must be true when needHandling(AsyncEvents) is true.

The fix introduces updateThreadStopRequestIfNeeded(), which synchronizes
m_threadStopRequested to match m_trapBits regardless of how we got to the
current state. Instead of having separate requestThreadStopIfNeeded() and
cancelThreadStopIfNeeded() calls that must reason about complex interleavings,
we have a single function that ensures the flag matches the source of truth.

Key changes:
- fireTrap() and clearTrap() now call updateThreadStopRequestIfNeeded() after
  modifying trap bits (with compiler fence to prevent reordering).
- updateThreadStopRequestIfNeeded() checks if m_threadStopRequested matches
  needHandling(AsyncEvents) and calls the appropriate helper if out of sync.
- requestThreadStopIfNeeded() and cancelThreadStopIfNeeded() are simplified
  to just perform their work, with asserts instead of guards.
- All interleavings now converge to the correct state because we sync to
  the source of truth rather than trying to maintain invariants.

This approach is conceptually simpler: instead of reasoning about whether to
request or cancel based on what operation we're performing, we simply ensure
the derived state (m_threadStopRequested) matches the source of truth
(m_trapBits). The lock-free fast path for trap bit modifications is preserved.

Tests: The corresponding stress tests will be attached in a follow-up patch
Canonical link: https://commits.webkit.org/303185@main
